### PR TITLE
Add Vector Tile Service

### LINKED ISSUE #5303: [feature] add vector tile page to the Datasource Manager
Original commit: https://github.com/qgis/QGIS/commit/f786e5d75e4da74fb6ed8614d912b77510cb7086 by alexbruy

Unfortunately this naughty coder did not write a description... :-(

### LINKED ISSUE #5327: [FEATURE] Vector tiles writer API
Original commit: https://github.com/qgis/QGIS/commit/41db54aa16f65f9a3c8ec4964844d14151541c43 by wonder-sk

Initial work... Currently supporting output to a directory based on XYZ template, using Mapbox vector tiles encoding.

New classes:
- QgsVectorTileMVTEncoder - low-level class that operates on a single tile, converts vector features to raw tile byte array, for internal use
- QgsVectorTileWriter - higher level class that manages generation of multiple tiles, for use by clients
- QgsVectorTileMVTUtils - assorted helper functions for MVT encoding/decoding

### LINKED ISSUE #6079: [feature][vectortiles] Add option to create connections for ArcGIS Vector
Tile Services to easily load vector tile layers from these servers
Original commit: https://github.com/qgis/QGIS/commit/9114e91ff62aea03d9e1f720994ef3d4a10fc1e5 by nyalldawson

Adds new options in the Vector Tiles browser tree context menu
and Data Source Manager to create a connection to an ESRI VectorTileServer.
When loading the layer from the connection, the tile server parameters
will be automatically setup to suit the layer source.

### LINKED ISSUE #6085: [feature][vectortiles] Add checkbox for showing only visible rules
in the vector tiles labeling and renderer widgets
Original commit: https://github.com/qgis/QGIS/commit/92c3a88da4254b6bbf59a7bd0f2b869d7cb1ecf0 by nyalldawson

Makes it much easier to locate troublesome rules or work with
complex vector tile styling

### PR DESCRIPTION
**Pull request by lajunek**
Goal:Users shall be able to create a connection to a Vector Tile service.

Ticket(s): fix #5303 fix #6079 fix #5327

- [ ] Backport to LTR documentation is required

### CODE REVIEW

**Comment by DelazJ** [comment]
@lajunek afaics #6085 is not at this level of feature. It's found inside the vector tile layer properties, probably symbology tab. Vector tile probably requires a new chapter for exposing properties, the same as mesh/raster/vector, to fix #6085.
From a very quick glance (based on title), one of #5303 and #5215 would be a better candidate.

**Review by DelazJ — COMMENTED** [review]
Some more feedback. Thanks @lajunek

**Comment by lajunek** [comment]
@DelazJ How do you think about this new description?

**Comment by DelazJ** [comment]
Looks better to me @lajunek. Thanks.
Maybe should we mention the min and max scales?

File: docs/user_manual/managing_data_source/opening_data.rst [hunk 1]
```diff
@@ -1153,6 +1153,9 @@ You can add a service by opening the context menu with a right-click
 and choosing :guilabel:`New Connection`.
 You set up a service by adding a :guilabel:`Name` and a :guilabel:`URL`.
 The Vector Tiles Service must provide tiles in :file:`.pbf` format.
+The dialog provides two menus to define the :checkbox:`Min. Zoom Level`
```
Comment (DelazJ) [inline docs/user_manual/managing_data_source/opening_data.rst:1156]:
```suggestion
The dialog provides two menus to define the :guilabel:`Min. Zoom Level` 
```

**Review by DelazJ — COMMENTED** [review]
Thanks Larissa.

@wonder-sk would you mind giving a look? For people not used to that type of data it's not obvious, and I can't find an issue report that explains the matters. Eg, one question is whether pbf is the only supported format (in the case there exists more), what do represent the zoom levels values... Thanks.

File: docs/user_manual/managing_data_source/opening_data.rst [hunk 2]
```diff
@@ -1141,6 +1141,22 @@ WFS(-T), WCS, CSW, ...).
...
+---------------------------
+
+Vector Tiles services can be found in the :guilabel:`Vector Tiles`
+top level entry in the :guilabel:`Browser`.
+You can add a service by opening the context menu with a right-click
+and choosing :guilabel:`New Connection`.
+You set up a service by adding a :guilabel:`Name` and a :guilabel:`URL`.
+The Vector Tiles Service must provide tiles in :file:`.pbf` format.
+The dialog provides two menus to define the :checkbox:`Min. Zoom Level` 
+and the :checkbox:`Max. Zoom Level` which are dependend from the
```
[the remark this answers is not in the document]
Comment (lajunek) [reply docs/user_manual/managing_data_source/opening_data.rst]: @DelazJ I agree. We need to clarify your questions.

File: docs/user_manual/managing_data_source/opening_data.rst [hunk 3]
```diff
@@ -1141,6 +1141,25 @@ WFS(-T), WCS, CSW, ...).
...
+
+Using Vector Tiles services
+---------------------------
+
+Vector Tiles services can be found in the :guilabel:`Vector Tiles`
+top level entry in the :guilabel:`Browser`.
+You can add a service by opening the context menu with a right-click
+and choosing :guilabel:`New Connection`.
+You set up a service by adding a :guilabel:`Name` and a :guilabel:`URL`.
+The Vector Tiles Service must provide tiles in :file:`.pbf` format.
```
Comment (havatv) [inline docs/user_manual/managing_data_source/opening_data.rst:1155]: The OSM or Mapbox pbuf format / implementation?

Comment (lajunek) [reply docs/user_manual/managing_data_source/opening_data.rst:1155]: @wonder-sk can we only load Vector Tiles with the .pbf format?

Comment (lajunek) [reply docs/user_manual/managing_data_source/opening_data.rst:1155]: @DelazJ @havatv the idea is to mention the ArcGIS Map Tiles server as an example connection. What are your votes?

Comment (DelazJ) [reply docs/user_manual/managing_data_source/opening_data.rst:1155]: @lajunek, do you suggest to add a screenshot showing an ArcGIS service (in replacement of mapTiler?) or do I misunderstand what you are talking about? If case 1, I don't think it's necessary. I'm fine with what is already done.

**Comment by lajunek** [comment]
@DelazJ I'd like to merge and insert a different screenshot with a new PullRequest. We need a screenshot depicting the 3.16 GUI.

File: docs/user_manual/managing_data_source/opening_data.rst [hunk 4]
```diff
@@ -1141,6 +1141,40 @@ WFS(-T), WCS, CSW, ...).
...
+The dialog provides two menus to define the 
+|checkbox|:guilabel:`Min. Zoom Level` and the
+|checkbox|:guilabel:`Max. Zoom Level`. Vector Tiles have a 
+pyramid structure. By using these options you have the opportunity 
+to individually generate layers from the tile pyramid. These layers
+will then be used to render the Vector Tile in QGIS.
+For Mercator projection (used by OpenStreetMap Vector Tiles) Zoom Level 0
+represents the whole world at a scale of 1:500.000.000. Zoom Level 14
+represents the scale 1:35.000. 
+figure_vector_tiles_maptilerplanet_ shows the dialog with the
```
Comment (havatv) [inline docs/user_manual/managing_data_source/opening_data.rst:1165]:
```suggestion
:numref:`figure_vector_tiles_maptilerplanet` shows the dialog with the

```

File: docs/user_manual/managing_data_source/opening_data.rst [hunk 5]
```diff
@@ -1141,6 +1141,40 @@ WFS(-T), WCS, CSW, ...).
...
+
+.. _figure_vector_tiles_maptilerplanet:
+
+.. figure:: img/vector_tiles_maptilerplanet.png
+   :align: center
+   
+   Vector Tiles - Maptiler Planet configuration
+
+By using :guilabel:`New ArcGIS Vector Tile Service Connection ...` you can
+connect to ArcGIS Vector Tile Services.
```
Comment (havatv) [inline docs/user_manual/managing_data_source/opening_data.rst:1176]:
```suggestion
connect to ArcGIS Vector Tile Services (:numref:`figure_vector_tiles_maptilerplanet`).
```

**Review by havatv — COMMENTED** [review]
Use `numref` for better figure referencing

**Comment by lajunek** [comment]
@havatv do I need travis to be completed to merge the PR?

**Comment by havatv** [comment]
> @havatv do I need travis to be completed to merge the PR?

I don't think so (it is probably unrelated to this PR).

File: docs/user_manual/managing_data_source/opening_data.rst (hunk 5 above)
Comment (DelazJ) [reply docs/user_manual/managing_data_source/opening_data.rst:1175]: I can't see any reference to the arcgis service in the screenshot so I'm not convinced we need it here

**Comment by DelazJ** [comment]
> @DelazJ I'd like to merge 

No objection though I'm surprised to see that the recent commits brought back changes that were removed previously (eg, mention of map tile planetservice). And unfortunately I do not expect reply for the diverse questions you/we raised in the PR and the list about this format/provider, so feel free...
> and insert a different screenshot with a new PullRequest.  We need a screenshot depicting the 3.16 GUI.

They look the same to me already, don't they?

**Comment by lajunek** [comment]
@havatv can somebody help me with merging this PR?

File: docs/user_manual/managing_data_source/opening_data.rst (hunk 5 above)
Comment (lajunek) [reply docs/user_manual/managing_data_source/opening_data.rst:1175]: @ me too, the screenshot shows the connection for maptiler planet data

**Pull request merged**